pkg/ingester/ingester.go

```diff
@@ -2649,3 +2685,41 @@ func (i *Ingester) UserRegistryHandler(w http.ResponseWriter, r *http.Request) {
 		Timeout:            10 * time.Second,
 	}).ServeHTTP(w, r)
 }
+
+// findStorageLabelMatcher returns value of storage label matcher and its index, if it exists.
+func findStorageLabelMatcher(matchers []*labels.Matcher) (string, int, error) {
+	resultVal, resultIdx := "", -1
+
+	for idx, matcher := range matchers {
+		if matcher.Name == StorageLabelName {
+			if resultIdx >= 0 {
+				return "", idx, errMultipleStorageMatchersFound
+			}
+			if matcher.Type != labels.MatchEqual {
+				return "", idx, errInvalidStorageMatcherType
+			}
+			resultVal = matcher.Value
+			resultIdx = idx
+		}
+	}
+
+	return resultVal, resultIdx, nil
+}
+
+// removeStorageMatcher returns the value of storage label (or empty string, if not found), and input matchers without the storage label matcher.
+func removeStorageMatcher(matchers []*labels.Matcher) (val string, filtered []*labels.Matcher, _ error) {
+	val, idx, err := findStorageLabelMatcher(matchers)
+	if err != nil {
+		return "", matchers, err
+	}
+	if idx < 0 {
+		return "", matchers, nil
+	}
+
+	// Create a new slice with the shard matcher removed.
+	filtered = make([]*labels.Matcher, 0, len(matchers)-1)
+	filtered = append(filtered, matchers[:idx]...)
+	filtered = append(filtered, matchers[idx+1:]...)
```

Comment:
why do we need to create a new slice, instead of just modifying the existing one?

Reply:
Keep in mind you will not even notice it among ingester allocations.

Reply:
Updated in https://github.com/grafana/mimir/pull/3961/commits/8243400a9dd286aea67ee132c079387a6589472f.

Reply:
While at it, we can avoid copy if it's the first index :) Yay for micro-optimizations.